jwst/extract_1d/soss_extract/atoca.py

```diff
@@ -1277,7 +1278,14 @@ def _solve(matrix, result):
         # Only solve for valid indices, i.e. wavelengths that are
         # covered by the pixels on the detector.
         # It will be a singular matrix otherwise.
-        sln[idx] = spsolve(matrix[idx, :][:, idx], result[idx])
+        with warnings.catch_warnings(action="error"):
```

Comment:
Unfortunately the `catch_warnings` context manager didn't gain the `action` parameter until py 3.11, and we support 3.9+. Suggestion:

```
    warnings.filterwarnings(action='error', category=MatrixRankWarning)
    try:
        sln[idx] = spsolve(matrix[idx, :][:, idx], result[idx])
    except MatrixRankWarning:
        # on rare occasions spsolve's approximation of the matrix is not appropriate
        # and fails on good input data. revert to different solver
        log.info('ATOCA matrix solve failed with spsolve. Retrying with least-squares.')
        sln[idx] = lsqr(matrix[idx, :][:, idx], result[idx])[0]
    warnings.resetwarnings()
```

Reply:
I confirmed this failure locally, and confirmed that your proposed fix works on both 3.9 and 3.11 locally.